examples/pytorch/graphsage/node_classification.py

```diff
@@ -35,98 +30,113 @@ def forward(self, blocks, x):
-        y = blocks[-1].dstdata['label']
-        y_hat = model(blocks, x)
-        loss = F.cross_entropy(y_hat, y)
-        opt.zero_grad()
-        loss.backward()
-        opt.step()
-        if it % 20 == 0:
-            acc = MF.accuracy(y_hat, y)
-            mem = torch.cuda.max_memory_allocated() / 1000000
-            print('Loss', loss.item(), 'Acc', acc.item(), 'GPU Mem', mem, 'MB')
-    tt = time.time()
-    print(tt - t0)
-    durations.append(tt - t0)
-
+    
+def evaluate(model, graph, dataloader):
     model.eval()
     ys = []
     y_hats = []
-    for it, (input_nodes, output_nodes, blocks) in enumerate(valid_dataloader):
+    for it, (input_nodes, output_nodes, blocks) in enumerate(dataloader):
         with torch.no_grad():
             x = blocks[0].srcdata['feat']
             ys.append(blocks[-1].dstdata['label'])
             y_hats.append(model(blocks, x))
-    acc = MF.accuracy(torch.cat(y_hats), torch.cat(ys))
-    print('Validation acc:', acc.item())
+    return MF.accuracy(torch.cat(y_hats), torch.cat(ys))
+
+def layerwise_infer(args, device, graph, nid, model, batch_size):
+    model.eval()
+    with torch.no_grad():
+        pred = model.inference(graph, device, batch_size)
+        pred = pred[nid]
+        label = graph.ndata['label'][nid]
+        return MF.accuracy(pred, label)
+
+def train(args, device, g, dataset, model):
+    # create sampler & dataloader
+    train_idx = dataset.train_idx.to(device)
+    val_idx = dataset.val_idx.to(device)
+    sampler = NeighborSampler([15, 10, 5],
+                              prefetch_node_feats=['feat'],
+                              prefetch_labels=['label'])
+    use_uva = (args.mode == 'mixed')
+    train_dataloader = DataLoader(g, train_idx, sampler, device=device,
+                                  batch_size=1024, shuffle=True,
+                                  drop_last=False, num_workers=0,
+                                  use_uva=use_uva)
+
+    val_dataloader = DataLoader(g, val_idx, sampler, device=device,
+                                batch_size=1024, shuffle=True,
+                                drop_last=False, num_workers=0,
+                                use_uva=use_uva)
+
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
+    
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (input_nodes, output_nodes, blocks) in enumerate(train_dataloader):
+            x = blocks[0].srcdata['feat']
+            y = blocks[-1].dstdata['label']
+            y_hat = model(blocks, x)
+            loss = F.cross_entropy(y_hat, y)
+            opt.zero_grad()
+            loss.backward()
+            opt.step()
+            total_loss += loss.item()
+        acc = evaluate(model, g, val_dataloader)
+        print("Epoch {:05d} | Loss {:.4f} | Accuracy {:.4f} "
+              .format(epoch, total_loss / (it+1),
+                                          acc.item()))
```

Comment:
Why do we need a new line for it?

Reply:
We don't. It might be due to my text editor or maybe I forgot it when deleting GPU mem.... Fixed. Thanks for catching it.